### ISSUE TITLE
Pod.Status.PodIP not updated during postStart lifecycle hook

### ISSUE THREAD

**Issue by spikecurtis** [kubernetes/kubernetes#85966]
**What happened**:

When using a postStart lifecycle hook, the Pod.Status.PodIP is not updated while the hook is running, even though the pod has been assigned an IP via the CNI plugin and networking is set up.

**What you expected to happen**:

The Pod.Status.PodIP should be set as soon as it's available.

**How to reproduce it (as minimally and precisely as possible)**:

The following manifest should repro the issue. Apply it and then watch the pod status.

```
apiVersion: v1
kind: Pod
metadata:
  name: test-calico-post-start
spec:
  containers:
  - name: test-calico
    image: byrnedo/alpine-curl:latest
    command:
    - sh
    - -c
    - "sleep 400000"
    lifecycle:
      postStart:
        exec:
          command: [ "/bin/sh","-c","sleep 120 && result=$(curl  http://www.google.com 1> /var/log/output.log 2> /var/log/output-error.log ); echo $result"]
```

The postStart lifecycle hook here contains a sleep for 120 seconds. During the first 120 seconds, the pod status will be ContainerCreating and the podIP will not be set. Only after the hook completes will the podIP be set.

**Anything else we need to know?**:

Why is this important? It prevents Calico network policy from operating correctly during the postStart hook since we cannot learn the pod's IP address. (Note that when Calico is also the CNI plug-in, we do learn it, but Calico is designed to run in "policy only mode" on top of other CNI plug-ins, like the AWS VPC-CNI plug-in). C.f. https://github.com/projectcalico/libcalico-go/issues/1125 and https://github.com/projectcalico/felix/issues/2008 for users who are having problems here.

**Environment**:
- Kubernetes version (use `kubectl version`): v1.16.3
- Cloud provider or hardware configuration: AWS
- OS (e.g: `cat /etc/os-release`): Ubuntu 16.04
- Kernel (e.g. `uname -a`): 4.4.0-1098-aws
- Install tools: kubeadm
- Network plugin and version (if this is a network-related bug): amazon-vpc-cni-k8s v1.5 with Calico v3.8.1
- Others:

**Comment by spikecurtis** [comment]
@kubernetes/sig-node-bugs

**Comment by spikecurtis** [comment]
Poking through the code a bit, it looks like exec-style post-start hooks run *synchronously* (and in series) within the `kubeGenericRuntimeManager`'s `SyncPod` function. This is the same function that creates the sandbox, so it seems possible that kubelet creates the sandbox (assigning pod IP), then starts the container, and executes the blocking hook, which prevents the `SyncPod` from completing and any updates from Kubelet from being pushed to API server.

Does that sound right?

**Comment by tedyu** [comment]
I took a look at pkg/kubelet/kuberuntime/kuberuntime_manager.go

I didn't seem to find where post-start hook is called.

**Comment by xiaoanyunfei** [comment]
> Poking through the code a bit, it looks like exec-style post-start hooks run _synchronously_ (and in series) within the `kubeGenericRuntimeManager`'s `SyncPod` function. This is the same function that creates the sandbox, so it seems possible that kubelet creates the sandbox (assigning pod IP), then starts the container, and executes the blocking hook, which prevents the `SyncPod` from completing and any updates from Kubelet from being pushed to API server.
> 
> Does that sound right?

That is exactly right

Pod status should be `ContainerCreating` until `postStart` is finished, and  pod IP could be updated during next `SyncPod`.
Maybe we can update pod IP(exclude pod status)  just before  `postStart`
But the logic seem be a little strange

**Comment by tedyu** [comment]
How about making the PostStart call asynchronously and letting startContainer() return a channel (of error) so that the caller can check whether the PostStart hook succeeds or not ?

**Comment by tedyu** [comment]
There is a simple change where we pass environment variables to the hook:
```
diff --git a/pkg/kubelet/kuberuntime/kuberuntime_container.go b/pkg/kubelet/kuberuntime/kuberuntime_container.go
index 20a35899f28..ff3905a96fd 100644
--- a/pkg/kubelet/kuberuntime/kuberuntime_container.go
+++ b/pkg/kubelet/kuberuntime/kuberuntime_container.go
@@ -180,7 +180,7 @@ func (m *kubeGenericRuntimeManager) startContainer(podSandboxID string, podSandb
                        Type: m.runtimeName,
                        ID:   containerID,
                }
-               msg, handlerErr := m.runner.Run(kubeContainerID, pod, container, container.Lifecycle.PostStart)
+               msg, handlerErr := m.runner.Run(kubeContainerID, pod, container, container.Lifecycle.PostStart, containerConfig.Envs)
                if handlerErr != nil {
                        m.recordContainerEvent(pod, container, kubeContainerID.ID, v1.EventTypeWarning, events.FailedPostStartHook, msg)
                        if err := m.killContainer(pod, kubeContainerID, container.Name, "FailedPostStartHook", nil); err != nil {
```
environment variables have the Pod IP.

**Comment by spikecurtis** [comment]
@tedyu the problem isn't that the hook doesn't have the podIP, it's that querying the pod on the APIServer doesn't have the podIP.

Calico talks to the APIServer to learn the podIP, and the problem is that during the hook, kubelet is blocked from updating the APIServer, so Calico is also blocked from setting up network policy, so the hook target can't talk on the network.  The hook itself isn't used by Calico --- it's end users that want to use the hook, and they want to talk on the network with their hook target.

**Comment by spikecurtis** [comment]
Either letting the hooks run async or pushing an APIServer pod status update immediately following sandbox creation (that sets the pod IP, but leaves the state ContainerCreating) would work for our use case.

**Comment by tedyu** [comment]
/cc @dchen1107 @derekwaynecarr @yujuhong 

Which option from above is preferred ?

Thanks

**Comment by spikecurtis** [comment]
I don't have a strong preference, but in the case of multiple containers one or more of which have long-running hook targets, it would be nice to allow them to run async and see pod status updates that track the progress of each container.  Right now if you have 5 containers and the first one blocks, none of the other containers start and you don't see any APIServer updates until everything is done.

**Comment by tedyu** [comment]
The PR makes the post start hook execution async by introducing a bool parameter to startContainer.
When the parameter is true, post start hook execution would be async.
When it is false, post start hook execution would be synchronous. This is used by init containers.

**Comment by spikecurtis** [comment]
If I understand correctly, this means that init containers with postStart hooks could still block, right? Why is that desirable?

I think that means that if an end user wants to have an init container with a postStart hook that talks on the network, they are going to still have the problem.

**Comment by tedyu** [comment]
This is the current code:
```
		if err := start("init container", container, false); err != nil {
			return
```
I am just keeping backward compatibility.

**Comment by spikecurtis** [comment]
Could you elaborate on what you mean by backward compatibility? Why do init containers' hooks need to be executed synchronously but it's OK for regular containers to be async?

**Comment by tedyu** [comment]
In order to prevent init container start from hanging, we can put the call to post start hook in a goroutine whose execution time is bounded (2 minutes, e.g.)
If the execution time exceeds the limit, return timeout error for init container start.

**Comment by xiaoanyunfei** [comment]
https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#hook-handler-execution
> Hook handler calls are synchronous within the context of the Pod containing the Container. This means that for a PostStart hook, the Container ENTRYPOINT and hook fire asynchronously. However, if the hook takes too long to run or hangs, the Container cannot reach a running state.

@tedyu 
Won't this violate the file？

**Comment by haosdent** [comment]
would it still possible that didn't get pod IP even change to async? For example, the async block run before `SyncPod`?

**Comment by tedyu** [comment]
@xiaoanyunfei
Not sure if you have looked at the PR.
The current code doesn't check error return from container start. There is no change in this regard.

If the hook hangs, the container wouldn't reach running state (with PR).

**Comment by xiaoanyunfei** [comment]
> @xiaoanyunfei
> Not sure if you have looked at the PR.
> The current code doesn't check error return from container start. There is no change in this regard.

I'm a little confused
kubelet actually check the return `msg, err` from `startContainer`, is there any doubt on this ?

**Comment by tedyu** [comment]
The return value is not checked by SyncPod() for ephemeral and normal containers.
See line 825:
```
			start("ephemeral container", c, true)
```
and line 835:
```
		start("container", &pod.Spec.Containers[idx], true)
```

**Comment by spikecurtis** [comment]
> In order to prevent init container start from hanging, we can put the call to post start hook (w.r.t. init container) in a goroutine whose execution time is bounded (2 minutes, e.g.)
> If the execution time exceeds the limit, return timeout error for init container start.

Does `SyncPod()` still block waiting for the hook to complete or timeout to fire?  If so, then this doesn't fully solve our problem, because we will still be blocked from getting the podIP while that hook is executing.

If we need to keep the init container hooks synchronous, then I would prefer the solution where we trigger a pod status update including the pod IP immediately following sandbox creation and prior to going into any blocking code starting containers.

**Comment by spikecurtis** [comment]
@tedyu @xiaoanyunfei I'd really like to settle on an approach here, even if no one is able to write up a fix yet.

I think we've got two options being discussed:
 1) Hooks run async of the goroutine that runs `SyncPod()`. In order for this to fully solve the problem, *all* hooks would need to run async, including those for init containers.
2) Trigger a pod status update immediately following sandbox creation, not blocked by any container creation of hooks.

WDYT?

**Comment by tedyu** [comment]
For option #2, we can break kubeGenericRuntimeManager#SyncPod into two parts.
SyncPodSandbox performs steps 1 thru 4 of the original:
```
-func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, podStatus *kubecontainer.PodStatus, pullSecrets []v1.Secret, backOff *flowcontrol.Backoff) (result kubecontainer.PodSyncResult) {
+func (m *kubeGenericRuntimeManager) SyncPodSandbox(pod *v1.Pod, podStatus *kubecontainer.PodStatus, pullSecrets []v1.Secret, backOff *flowcontrol.Backoff) (result kubecontainer.PodSyncResult, podIPs []string, podContainerChanges podActions) {
```
SyncPod performs 5, 6 and 7
```
+//  5. Create ephemeral containers.
+//  6. Create init containers.
+//  7. Create normal containers.
+func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, podStatus *kubecontainer.PodStatus, pullSecrets []v1.Secret, backOff *flowcontrol.Backoff, podIPs []string, podSandboxID string, podContainerChanges podActions) (result kubecontainer.PodSyncResult) {
```
so that kubelet can call the following in between these two funcs:
```
	// Update status in the status manager
	kl.statusManager.SetPodStatus(pod, apiPodStatus)
```
However, podActions needs to be passed between the two funcs.

@derekwaynecarr @dchen1107 @yujuhong 
Do you have better idea ?

**Comment by tedyu** [comment]
I have implemented the above approach in #86271

**Comment by tedyu** [comment]
@spikecurtis 
Can you bring this feature request up at the next sig-node meeting ?

**Comment by Random-Liu** [comment]
@spikecurtis Thanks for filing the issue!

One question: What if I have a pod with 10 containers, and the last one takes 2 min to start. In that case, even without the post start hook, we'll have similar issue, right? Do we expect that also to work?

**Comment by spikecurtis** [comment]
@Random-Liu Without the post-start hook, I believe that things work fine.  When Kubelet starts containers it doesn't wait for them to complete before sending a pod status update. So, the container may not be able to talk on the network the instant it starts, but eventually Calico gets the pod status update and is able to set up the networking. Calico is designed to work in this "eventually consistent" manner, and this allows us to achieve high scale. In practice the time is very small and not noticeable.

However, with the post-start hooks, Kubelet is blocked from sending the update until the hook completes its execution, and this causes the problem.

**Comment by tedyu** [comment]
If we make execution of postStart hooks for init containers asynchronous, #86010 would be a preferable choice (which is simpler compared to #86271).

**Comment by tedyu** [comment]
How do we handle post start hook failure which may take some time to be detected (when execution of postStart hooks for init containers is asynchronous) ?
Should the normal container(s) be torn down once the post start hook failure is seen ?

**Comment by spikecurtis** [comment]
@tedyu It's not clear to me what the _desired_ behavior is if a hook fails.  The Kubernetes docs only say: 

> If a handler fails for some reason, it broadcasts an event.

https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#hook-handler-execution

If broadcasting an event is all, I think we are fine.  If there is some other behavior associated with a failed hook, can someone point to a definitive spec of what it should be?

**Comment by tedyu** [comment]
ping @derekwaynecarr @dchen1107 @yujuhong 
See if you know the spec mentioned by @spikecurtis above.

**Comment by dashpole** [comment]
@spikecurtis I talked with @yujuhong after sig-node today.  The pod creation process can be difficult to make changes to correctly, as it is a fairly complex part of the code-base.  But we think we can eventually work out the details in a code review.

I plan to take a pass at @tedyu's PR, but we have one other option that may work, and may have other benefits as well that we would like you to consider.  Instead of getting the pod's IP from the API server in the calico daemonset, you could get it from the CRI instead: https://github.com/kubernetes/cri-api/blob/master/pkg/apis/runtime/v1alpha2/api.proto#L418.  I don't know enough about how calico works to know if it is a good solution or not, but we should at least explore it.  Part of me wonders if that would decrease latency for calico network policy enforcement...

**Comment by spikecurtis** [comment]
@dashpole thanks for the suggestion. It was mentioned on one of the calls before, and I would say it's definitely not our preferred solution.  

One issue is that we need the pod IP everywhere in the cluster, not just the node the pod is sitting on. This is because remote nodes are enforcing network policy. Imagine that a postStart hook needs to talk to some other service in the K8s cluster, and we have a network policy that protects that service on ingress. Until the remote node hosting the service learns the pod IP, it can't allow the traffic.

I'm also not clear on how the CRI would work in practice. Is this an API that kubelet presents?

How is access to this API controlled?  Looking over it, it's pretty powerful, including the ability to start and stop arbitrary containers and run arbitrary exec commands in them. The Calico daemonset is currently highly privileged, but it's on our roadmap to dial down how much power it has. I'm concerned adding a dependency on the CRI API would make this much harder unless there is fine-grained access control.

**Comment by tedyu** [comment]
There is another PR #88170 which also tries to solve this issue.

It seems only #88169 shows up as linked to this.

**Comment by dashpole** [comment]
> One issue is that we need the pod IP everywhere in the cluster, not just the node the pod is sitting on. This is because remote nodes are enforcing network policy. Imagine that a postStart hook needs to talk to some other service in the K8s cluster, and we have a network policy that protects that service on ingress. Until the remote node hosting the service learns the pod IP, it can't allow the traffic.

Makes sense.  We weren't sure if just the local node needed it, or if other nodes did as well.  Thanks for the clarification.

**Comment by jonashartwig** [comment]
Hi, I just came across the same problem (using weave and k8s v1.17.4). I used before 1.12.7 and wanted to migrate to the newer version. This was not an issue in 1.12.7.

regards

**Comment by tedyu** [comment]
@spikecurtis has provided comment on using the solution from #88169

https://github.com/kubernetes/kubernetes/pull/88169#issuecomment-600883890

**Comment by jonashartwig** [comment]
I see he uses a new kubelet build from master. I use rancher/rke to facilitate my cluster. I will have a look if i can use a newer version of kublet with it. Do I understand correctly that this could be a way forward?

**Comment by ehashman** [comment]
/sig network
/help

This should be clear enough for someone to pick up to try to address with reproducer in first comment. We have a lot of refactoring happening in the pod lifecycle right now, I am curious if this will still be an issue in 1.22.

**Comment by siegenthalerroger** [comment]
Has anyone tested this under 1.22 yet? I currently am waiting on AKS (non preview) access and will try once I can.

**Comment by siegenthalerroger** [comment]
/reopen 
/remove-lifecycle rotten

As mentioned above, I haven't been able to test this on a new 1.22 cluster as of yet, so I can't verify that this is actually fixed/done.

**Comment by danwinship** [comment]
> Why is this important? It prevents Calico network policy from operating correctly during the postStart hook since we cannot learn the pod's IP address. (Note that when Calico is also the CNI plug-in, we do learn it, but Calico is designed to run in "policy only mode" on top of other CNI plug-ins, like the AWS VPC-CNI plug-in).

Note that even when the CNI plugin and the NetworkPolicy implementation are the same network plugin, this is still potentially very complicated, because it may need to communicate the pod IP to parts of the implementation on other nodes. Eg, if a pod X on Node A tries to contact a pod Y on Node B as part of its `postStart`, this may require Node B to update policy rules to allow pod Y to accept a connection from pod X's IP.

Either kubelet needs to update `pod.Status.PodIP` right away after the CNI plugin returns, or else we need to make it official that pods _do not_ have network access during `postStart`.

**Comment by danibaeyens** [comment]
This issue is also creating some undesired effects on other projects like linkerd2 (i.e., https://github.com/linkerd/linkerd2/discussions/8458) when using calico for network policies, as it blocks the startup of some pods. 

linkerd2 on its latest version has added a `postStart` to ensure that the proxy container is ready before accepting traffic, but it cannot get ready because it's waiting the `postStart` to finish. Then, after 2 minutes, a container restarts, `Pod.Status.PodIP` gets updated and works on its second try.

I rebased https://github.com/kubernetes/kubernetes/pull/88169 during last week, as it seemed (2 years ago) the right way to solve the issue. I was about to submit a new PR, but I've just seen your last comment.

Do you think it is valuable to revive that PR? Or does it need a different approach? (even considering the clarification of not having network access during `postStart`)

**Comment by shaneutt** [comment]
We discussed this one in the SIG Network sync today: we're not sure that this is firmly something for sig-network

/remove-sig network

However, if anyone disagrees with that, or just generally needs help from us to move this forward please do feel free to put it back and ping us! :vulcan_salute:

**Comment by haircommander** [comment]
I find myself wondering if we're going to address this in SIG-node. pod lifecycle is brittle and this could end up being a major change.

Is this scenario vital for SIG-network? It is not very high on SIG-node's priority list

/sig network
/triage needs-information

**Comment by aojea** [comment]
sig network seems the right home

**Comment by aojea** [comment]
it seems it has a workaround though https://github.com/openshift/cluster-network-operator/pull/1503

**Comment by aojea** [comment]
/assign

I want to go deeper into this problem

**Comment by AnishShah** [comment]
sig-node CI meeting:

@aojea , is this still an issue?

**Comment by SergeyKanzhelev** [comment]
/remove-sig node

please reapply as needed

**Comment by aojea** [comment]
/close

this is WAI , consumers of PodIP depending on the Pod.status.PodIPs field will always have this problem as the status is reflected after the IP is assigned.

I encourage components that need this information to start using new technologies like [api: add support for reading Pod IPs containerd/nri#119](https://github.com/containerd/nri/pull/119) NRI that is available by default in containerd 2.0 and latest CRIO releases and allow to integrate seamless with the runtime to obtain this information without having to modify anything in the host

**Comment by aojea** [comment]
Example of network policy implementation that does not have this problem using NRI https://github.com/kubernetes-sigs/kube-network-policies/pull/157
